### PR TITLE
Fix memory explosion multi runs

### PR DESCRIPTION
**Pull request by WenzDaniel**
**What is the problem / what does the code in this PR do**
In this PR we rework the job submission in multi-runs to address:  https://github.com/AxFoundation/strax/issues/563. The new function will always just submit twice as many jobs as there are workers. As soon as a job finishes its job the result is loaded and new jobs are submitted if needed. In this way we ensure that garbage collection is called in time on each individual worker. 

In addition I changed the function such that the output is sorted by run_id which was previously not the case. In terms of performance no significant increase in computing time can be observed. 

To test the different behavior just call 

```
st.make(run_ids, data_type)
```

or 

```
st.get_array(run_ids, data_type)
```

for more than a single run_id and observe the memory usage in python. 

I also added some test to ensure a correct behavior (doing an off by one error during the job submission was quite easy).

### CODE REVIEW

**Comment by WenzDaniel** [comment]
I did not put the pbar in a with statement as I tried to avoid another level of indentation.

**Review by JoranAngevaare — COMMENTED** [review]
Thanks Daniel! I know you spend a lot of time and frustration fixing this!

Were you able to profile the memory usage difference?

File: tests/test_utils.py [hunk 1]
```diff
@@ -146,3 +151,46 @@ def test_keep_drop_columns(d):
...
+        """Setup context and make some subruns."""
+        self.tempdir = tempfile.mkdtemp()
+        self.context = strax.Context(storage=[strax.DataDirectory(self.tempdir,
+                                                                  provide_run_metadata=True,
+                                                                  readonly=False,
+                                                                  deep_scan=True)],
+                                     register=[Records, Peaks],
+                                     config={'bonus_area': 42, 'use_per_run_defaults': False},
+                                     )
+        self.run_ids = [str(r) for r in range(5)]
```
Comment (JoranAngevaare) [inline tests/test_utils.py:169]: Have you run with more runs? Might be worth to do just  to be sure, especially if the test is fast

File: tests/test_utils.py [hunk 2]
```diff
@@ -146,3 +151,46 @@ def test_keep_drop_columns(d):
         assert np.all(selected_data[c] == d[c])
     for c in columns[:1]:
         assert c not in selected_data.dtype.names
+
+
+class TestMultiRun(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply tests/test_utils.py:156]: Sure, in that case I would have to profile first which value actually make sense. But can add it.

File: tests/test_utils.py [hunk 3]
```diff
@@ -146,3 +151,46 @@ def test_keep_drop_columns(d):
...
+        self.run_ids = [str(r) for r in range(5)]
+
+        for run_id in self.run_ids:
+            self.context.make(run_id, 'records')
+
+    def test_make_multi_run(self):
+        self._test_get_array_multi_run()
+        self._test_get_array_multi_run(max_worker=2)
+        self.context.set_context_config({'allow_multiprocess': True})
+        self._test_get_array_multi_run(max_worker=2)
```
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply tests/test_utils.py:178]: Okay sure I can make it individual tests, probably better for traceback.

File: tests/test_utils.py (hunk 1 above)
Comment (WenzDaniel) [reply tests/test_utils.py:172]: Is done now in the memory test.

**Review by JoranAngevaare — APPROVED** [review]
Thanks, test look thorough and if I were to try this for many runs, it would probably look like this, thanks a lot again!
